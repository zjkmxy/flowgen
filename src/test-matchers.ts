import { execFileSync } from "child_process";

import chalk from "chalk";
import flow from "flow-bin";

import { beautify } from ".";

// eslint-disable-next-line @typescript-eslint/no-unused-vars,no-unused-vars
declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace jest {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars,no-unused-vars
    interface Matchers<R> {
      toBeValidFlowTypeDeclarations(): R;
    }
  }
}

expect.extend({
  async toBeValidFlowTypeDeclarations(source) {
    const beautifiedSource = await beautify(source);
    try {
      execFileSync(
        flow,
        ["check-contents", "--color=always", "--timeout=30"],
        {
          input: beautifiedSource,
          stdio: ["pipe", "pipe", "pipe"],
        },
      );
    } catch (err) {
      return {
        message: () =>
          `expected ${chalk.bold(
            beautifiedSource.trimEnd(),
          )} to be valid flow:\n${chalk.red(err.stdout)}`,
        pass: false,
      };
    }

    return {
      message: () =>
        `expected ${chalk.bold(
          beautifiedSource.trimEnd(),
        )} not to be valid flow`,
      pass: true,
    };
  },
});
