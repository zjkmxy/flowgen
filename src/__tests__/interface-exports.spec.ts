import { compiler, beautify } from "..";
import "../test-matchers";

it("should handle exported interfaces", async () => {
  const ts = `export interface UnaryFunction<T, R> {
    (source: T): R;
  }
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should handle exported interfaces within a module", async () => {
  const ts = `declare module "my-module" {
    export interface UnaryFunction<T, R> {
      (source: T): R;
    }
  }
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});
