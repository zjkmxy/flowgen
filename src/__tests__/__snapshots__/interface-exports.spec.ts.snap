// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle exported interfaces 1`] = `
"export interface UnaryFunction<T, R> {
  (source: T): R;
}
"
`;

exports[`should handle exported interfaces within a module 1`] = `
"declare module "my-module" {
  declare export interface UnaryFunction<T, R> {
    (source: T): R;
  }
}
"
`;
